kratos/sources/kratos_application.cpp

```diff
@@ -295,4 +304,60 @@ void KratosApplication::RegisterKratosCore() {
     // Register ConstitutiveLaw BaseClass
     KRATOS_REGISTER_CONSTITUTIVE_LAW("ConstitutiveLaw", mConstitutiveLaw);
 }
+
+template<class TComponentsContainer>
+void KratosApplication::DeregisterComponent(std::string const & rComponentName) {
+    auto path = std::string(rComponentName)+"."+mApplicationName;
+
+    // Remove only if the application has this type of components registered
+    if (Registry::HasItem(path)) {
+
+        // Generate a temporal list with all the keys to avoid invalidating the iterator (Convert this into a transform range when C++20 is available)
+        std::vector<std::string> keys;
+        std::transform(Registry::GetItem(path).cbegin(), Registry::GetItem(path).cend(), std::back_inserter(keys), [](auto & key){return std::string(key.first);});
+
+        for (auto & key : keys) {
+            auto cmpt_key = "components."+key;
+            auto type_key = path+"."+key;
+
+            // Remove from KratosComponents
+            KratosComponents<TComponentsContainer>::Remove(key);
+
+            // Remove from registry general component list
+            if (Registry::HasItem(cmpt_key)) {
+                Registry::RemoveItem(cmpt_key);
```

Comment:
This would be tricky if we are to use the same mechanism for variables, because, multiple applications can define the same variable. And deregistering one application should not remove the variable, if added from multiple apps.